KAFKA-6244: Dynamic update of log cleaner configuration

rajinisivaram opened the pull request:
Log cleaner config update as described in KIP-226. Config updates are handled by stopping cleaner threads and starting new ones with the new config. This keeps the code simple and ensures that if any of the threads had terminated earlier due to an exception, new ones are created to match the configured thread count.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rajinisivaram committed:
KAFKA-6244: Dynamic update of log cleaner configuration

rajinisivaram commented:
@hachikuji When you have some time, can you review please? Thank you!

rajinisivaram commented:
retest this please

hachikuji commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -144,6 +152,40 @@ class LogCleaner(val config: CleanerConfig,
     cleaners.foreach(_.shutdown())
```
Maybe I'm missing it, but don't we need to clear `cleaners`? Perhaps there should be a test assertion somewhere.

rajinisivaram replied:
Oops, fixed and added test

hachikuji commented on core/src/main/scala/kafka/log/LogCleaner.scala:
```diff
@@ -144,6 +152,40 @@ class LogCleaner(val config: CleanerConfig,
+  }
+
+  private def logCleanerConfig(configs: util.Map[String, _]): CleanerConfig = {
+    CleanerConfig(numThreads = configs.get(KafkaConfig.LogCleanerThreadsProp).asInstanceOf[Int],
```
It's kind of a pity that `Reconfigurable` has to work with the `Map` interface since it means all this nasty casting. Ideally we could just pass through the `KafkaConfig` object itself and use the nice fields. I am tempted to suggest that we should have a separate `BrokerReconfigurable` interface which allows us to use the nicer object. What do you think?

rajinisivaram replied:
I had thought initially that we should use the same interface internally as we expose externally. But when I came to the final PR for listener update, I ended up doing that with KafkaConfig rather than maps since it was too messy otherwise. So it makes sense to use the same interface here. Updated.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -882,6 +882,16 @@ object LogManager {
   val LogStartOffsetCheckpointFile = "log-start-offset-checkpoint"
   val ProducerIdExpirationCheckIntervalMs = 10 * 60 * 1000
 
+  val LogCleanerReconfigurableConfigs = Set(
```
Seems like it might be more natural to put this in the `LogCleaner` companion object.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@hachikuji Thanks for the review. Have addressed the comments. Also changed the test for invalid props. Whereever `Properties` is used in `DynamicBrokerConfig`, we only expect string values (read from ZooKeeper or from the string config value in AdminClient). So the test that was setting an integer value in `Properties`can never actually occur. So I have removed that and added a test with an invalid string for an int config.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -343,18 +355,44 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
+  }
+
+  private def processBrokerReconfigurable(reconfigurable: BrokerReconfigurable,
+                                    oldConfig: KafkaConfig,
```
nit: weird alignment

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -343,18 +355,44 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
   }
 }
+
+trait BrokerReconfigurable extends Reconfigurable {
```
The overrides below make me a little uncomfortable. I guess the only reason we need this to extend `Reconfigurable` is because of the `reconfigurables` collection? Do you think it would be unreasonable to use two separate collections?

rajinisivaram replied:
@hachikuji Makes sense, updated. Thanks.

hachikuji reviewed (commented):
Thanks for the updates. Just a couple more comments.

rajinisivaram committed:
Address review comments

hachikuji reviewed (approved):
Thanks for the updates. LGTM
